File: src/geometryService/cpp/stractNet/MsgStop.cpp
///////////////////////////////////////////////////////////
//  MsgStop.cpp
//  Implementation of the Class MsgStop
//  Created on:      06-Aug-2008 7:51:11 AM
//  Original author: david.h.loman
///////////////////////////////////////////////////////////
#include "MsgStop.h"

///////////////////////////////////////////////////////////////////////
/*
 *
 * xSTRUCTORS
 *
 */
///////////////////////////////////////////////////////////////////////
MsgStop::~MsgStop() {
}
MsgStop::MsgStop(std::string Name, std::set<int> msgTypes, std::List<StdMsg> inQ) {
	SNRoot(Name);
	this->MsgTypes = msgTypes;
	this->msgQ = inQ;

	this->addMsgType(StdMsgTypes::mtBroadcast);
	this->addMsgType(StdMsgTypes::mtHello);
}
MsgStop::MsgStop(std::string Name, std::List<StdMsg> inQ) {
	SNRoot(Name);
	this->MsgTypes = msgTypes;
	this->msgQ = inQ;

	this->addMsgType(StdMsgTypes::mtBroadcast);
	this->addMsgType(StdMsgTypes::mtHello);
}
///////////////////////////////////////////////////////////////////////
/*
 *
 * METHODS
 *
 */
///////////////////////////////////////////////////////////////////////
void MsgStop::addMsgType(int msgType) {
	//All msgtype additions should go through here
	//to aid in sync and repetition of code

	//Add any QA here
	this->MsgTypes.insert(msgType);
}

void MsgStop::addMsgTypes(std::set<int> msgTypes) {
	//Iterate over all the elements in the incoming set
	//and add them to the MsgTypes set.
	for (std::set<int>::iterator loopIter = msgTypes.begin(); loopIter
			!= msgTypes.end(); ++loopIter) {
		this->addMsgType(*loopIter);
	}
}

void MsgStop::addMsgTypes(int msgTypes[]) {
	int numElements = sizeof(msgTypes) / sizeof(int);
	for (int i = 0; i < numElements; ++i) {
		this->addMsgType(msgTypes[i]);
	}
}

std::set<int>& MsgStop::getMsgTypes() {
	return this->MsgTypes;
}

//int MsgStop::getMsgTypesAsArray() {
//	//init the array to the correct size
//	int myArray[this->MsgTypes.size()];
//	int cnt = 0;
//	std:set<int>::iterator it;
//
//	for ( it = this->MsgTypes.begin(); it != this->MsgTypes.end();++it) {
//		myArray[cnt] = *it;
//		cnt++;
//	}
//	return ;
//}

void MsgStop::remMsgType(int msgType) {
	this->MsgTypes.erase(mgType);
}

void MsgStop::sendToMS(StdMsg msg) {
	SNRoot::out("Sent StdMsg to " & this->getLocalMS().gON() & " with MsgID/Type: "
			& msg.getMsgID() & " / " + msg.getMsgType(), 2);
	SNRoot::out("\n" + msg.ElementsToString(), 3);
	msg.setLocalFrom(this);
	this->getLocalMS().AddMsgToQueue(msg);

}

void MsgStop::sendToMsgStop(StdMsg msg) {
	SNRoot::out("Received StdMsg from " & this->getLocalMS().gON() & " with MsgID/Type: " & msg.getMsgID()
			& " / " & msg.getMsgType(), 2);
	SNRoot::out("\n" & msg.ElementsToString(), 3);
	this->msgQ.add(msg);
}

File: src/iBME/superceded_GS/java/stractNet/src/stractNet/messaging/MsgStop.java
package stractNet.messaging;

/*
 * Meant to be an Aggragate class.
 */

import java.util.LinkedList;
import java.util.HashSet;
import stractNet.messaging.msgs.StdMsg;
import stractNet.messaging.msgs.StdMsgTypes;
import stractNet.SNRoot;
import java.util.Iterator;

public class MsgStop extends SNRoot {

	// This is a reference to the Queue for messages coming FROM the MS to the
	// destined MsgStop
	private LinkedList<StdMsg> msgQ = null;
	private HashSet<Integer> MsgTypes;

	public MsgStop(String Name, HashSet<Integer> msgTypes,
			LinkedList<StdMsg> inQ) {
		super(Name);
		this.MsgTypes = msgTypes;
		this.msgQ = inQ;
		
		this.addMsgType(StdMsgTypes.mtBroadcast);
		this.addMsgType(StdMsgTypes.mtHello);
	}
	
	

	public MsgStop(String Name, LinkedList<StdMsg> inQ) {
		this(Name, new HashSet<Integer>(), inQ);
	}

	public void sendToMsgStop(StdMsg msg) {
		synchronized (this.msgQ) {
			this.out("Recieved StdMsg from " + this.getLocalMS().gON() + " with MsgID/Type: " + msg.getMsgID()
					+ " / " + msg.getMsgType(), 2);
			this.out("\n" + msg.ElementsToString(), 3);
			this.msgQ.add(msg);
			this.msgQ.notify();			
		}
	}
	
	

	public void sendToMS(StdMsg msg) {
		this.out("Sent StdMsg to " + this.getLocalMS().gON() + " with MsgID/Type: " + msg.getMsgID()
				+ " / " + msg.getMsgType(), 2);
		this.out("\n" + msg.ElementsToString(), 3);
		msg.setLocalFrom(this);
		this.getLocalMS().AddMsgToQueue(msg);
	}
	
	public HashSet<Integer> getMsgTypesHashSet() {
		return MsgTypes;
	}

	public int[] getMsgTypesIntArray() {

		Iterator<Integer> it = this.MsgTypes.iterator();
		int[] mts = new int[this.MsgTypes.size()];

		for (int i = 0; i < this.MsgTypes.size(); i++) {
			mts[i] = it.next();
		}

		return mts;
	}

	public void addMsgType(int msgType) {
		MsgTypes.add(msgType);
	}

//	public void addMsgTypes(HashSet<Integer> msgTypes) {
//		MsgTypes.addAll(MsgTypes);
//	}

	public void addMsgTypes(int[] msgTypes) {
		for (int t : msgTypes) {
			this.addMsgType(t);
		}
	}

	public void remMsgType(int msgType) {
		MsgTypes.remove(msgType);
	}


}
